devito/types/tensor.py

```diff
@@ -0,0 +1,603 @@
+from collections import OrderedDict
+
+import sympy
+import numpy as np
+from sympy.core.sympify import converter as sympify_converter
+from sympy.core.decorators import call_highest_priority
+
+from devito.finite_differences import Differentiable
+from devito.logger import error
+from devito.types.basic import AbstractCachedTensor, Basic
+from devito.types.dense import Function, TimeFunction
+from devito.types.utils import NODE
+
+__all__ = ['TensorFunction', 'TensorTimeFunction', 'VectorFunction', 'VectorTimeFunction']
+
+
+class TensorFunction(AbstractCachedTensor, Differentiable):
+    """
+    Tensor valued Function represented as a Matrix.
+    Each component is a Function or TimeFunction.
+
+    A TensorFunction and the classes that inherit from it takes the same parameters as
+    a DiscreteFunction and additionally:
+
+    Parameters
+    ----------
+    symmetric : Bool, optional
+        Wether the tensor is symmetric or not. Defaults to True.
+    diagonal : Bool, optional
+        Wether the tensor is diagonal or not. Defaults to False.
+    staggered: Tuple of Dimension, optional
```

Comment:
tuple, not Tuple

check for this sort of things across all docstrings in this file